Readme.md
Wally
====================

Quick and dirty wallpaper changer for PCManFM (preferably Lubuntu) written in python.

Requirements
---------------------
* python
* pygtk
* PCManFM
* ^their requirements^

Icon
---------------------
From [Tango Desktop Project](http://tango.freedesktop.org/Tango_Desktop_Project).
Licensed under public domain.

License
---------------------
Copyright (C) 2013 paxman

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Changes
v0.1 alpha, 14.3.2013 -- Initial release.
15.3 -- Fix image loading; add license

src/WallyUtil.py
import string

#from http://code.activestate.com/recipes/82102-smart-pluralisation-english/
def Plural(num=1, text=''):
    if num == 1:
        # singular -- easy
        result = '%d %s' % (num, text)
    else:
        aberrant = {    'knife'     : 'knives',
                'self'        : 'selves',
                'elf'        : 'elves',
                'life'        : 'lives',
                'hoof'        : 'hooves',
                'leaf'        : 'leaves',
                'echo'        : 'echoes',
                'embargo'    : 'embargoes',
                'hero'        : 'heroes',
                'potato'    : 'potatoes',
                'tomato'    : 'tomatoes',
                'torpedo'    : 'torpedoes',
                'veto'        : 'vetoes',
                'child'        : 'children',
                'woman'        : 'women',
                'man'        : 'men',
                'person'    : 'people',
                'goose'        : 'geese',
                'mouse'        : 'mice',
                'barracks'    : 'barracks',
                'deer'        : 'deer',
                'nucleus'    : 'nuclei',
                'syllabus'    : 'syllabi',
                'focus'        : 'foci',
                'fungus'    : 'fungi',
                'cactus'    : 'cacti',
                'phenomenon'    : 'phenomena',
                'index'        : 'indices',
                'appendix'    : 'appendices',
                'criterion'    : 'criteria'
                }

        if aberrant.has_key(text):
            result = '%d %s' % (num, aberrant[text])
        else:
            postfix = 's'
            if len(text) > 2:
                vowels = 'aeiou'
                if text[-2:] in ('ch', 'sh'):
                    postfix = 'es'
                elif text[-1:] == 'y':
                    if (text[-2:-1] in vowels) or (text[0] in string.uppercase):
                        postfix = 's'
                    else:
                        postfix = 'ies'
                        text = text[:-1]
                elif text[-2:] == 'is':
                    postfix = 'es'
                    text = text[:-2]
                elif text[-1:] in ('s', 'z', 'x'):
                    postfix = 'es'
                
            result = '%d %s%s' % (num, text, postfix)
    return result



src/WallyTimer.py
from gobject import timeout_add_seconds, source_remove

class WallyTimer():
    def __init__(self, manager):
        self.manager = manager
        self.interval = manager.config.get_selected_interval()
        self.timer = None 
    
    def stop(self):
        #if timer running
        if self.timer:
            source_remove(self.timer)
            self.timer = None
            print "Timer stopped."
            
        print "Timer already stopped."
    
    def set_interval(self, interval):
        #print interval
        self.interval = interval*60
        
        #if running
        if self.timer:
            self.stop()
            self.start()
            print "Timer restarted"
    
    def start(self):
        self.stop()
        self.set_interval(self.manager.config.get_selected_interval())
        self.timer = timeout_add_seconds(self.interval,self.manager.change_wallpaper)
        print "Timer started with interval of",str(self.interval),"seconds"

src/WallyConfig.py
import ConfigParser
import os
from WallyWrapper import WallyWrapper

class WallyConfig():
    
    def __init__(self):
        self.load()
        
    def load(self):
        user_path = os.path.expanduser("~")
        
        #use defaults 
        if not os.path.isfile(user_path+"/.config/wally/config.cfg"):
            self.set_default()
        else:
            try:
                config = ConfigParser.RawConfigParser(allow_no_value=True)
                config.read(user_path+"/.config/wally/config.cfg")
                self.interval = config.getint("wally", "interval")
                self.custom_interval = config.getint("wally", "custom_interval")
                self.interval_type = config.getint("wally", "interval_type")
                self.mode = config.get("wally", "mode")
                self.path = config.get("wally", "path")
            #if any errors
            except (ConfigParser.NoOptionError, ConfigParser.NoSectionError):
                self.set_default()
    
    def get_mode(self):
        return self.mode
    
    def get_interval(self):
        return self.interval
 
    def get_custom_interval(self):
        return self.custom_interval
    
    def get_selected_interval(self):
        if self.get_interval_type() == 0:
            return self.get_interval()
        else:
            return self.get_custom_interval()
    
    def get_interval_type(self):
        return self.interval_type
     
    def get_path(self):
        return self.path
    
    def set_mode(self, mode):
        self.mode = mode
    
    def set_interval(self, interval):
        self.interval = interval
        
    def set_custom_interval(self, custom_interval):
        self.custom_interval = custom_interval
    
    def set_interval_type(self, interval_type):
        self.interval_type = interval_type
            
    def set_path(self, path):
        self.path = path
    
    def set_default(self):
        print "Loading defaults.."
        self.interval = 5
        self.custom_interval = 6
        self.interval_type = 0
        self.mode = "fit"
        self.path = WallyWrapper.user_pictures_folder()
    
    def save_config(self):
        print "Saving..."
        #TODO: allow no value for excluded files 
        config = ConfigParser.RawConfigParser(allow_no_value=True)
        config.add_section("wally")
        config.set("wally",'interval', self.get_interval())
        config.set("wally",'custom_interval', self.get_custom_interval())
        config.set("wally",'interval_type', self.get_interval_type())
        config.set("wally",'mode', self.get_mode())
        config.set("wally",'path', self.get_path())
        user_path = os.path.expanduser("~")
        
        #create folder if it does not exist
        if not os.path.isdir(user_path+"/.config/wally/"):
            os.makedirs(user_path+"/.config/wally/")
            
        with open(user_path + "/.config/wally/config.cfg", 'wb') as configfile:
            config.write(configfile)
        print "...done"

src/WallyManager.py
from WallyConfig import WallyConfig
from WallyTimer import WallyTimer
from WallyFileRoller import WallyFileRoller
from WallyWrapper import WallyWrapper
from WallyGUI import WallyTray
from WallyGUI import WallyAboutDialog, WallyFolderSelector, WallyIntervalInput
import gtk,os

class WallyManager():
    
    default_intervals = (60,30,15,10,5,3,2,1)
    default_max_interval = 600
    
    def __init__(self):
        self.config = WallyConfig()
        self.wrapper = WallyWrapper()
        self.timer = WallyTimer(self)
        self.fileroller = WallyFileRoller(self.config.get_path())
        self.tray = WallyTray(self)
         
    def user_change_wallpaper(self, widget, event,**kwarg):
        if event.button == 1 and event.type == gtk.gdk._2BUTTON_PRESS:
            self.stop()
            self.change_wallpaper()
            self.start()
        
    def change_wallpaper(self):
        selected_file = self.fileroller.get_next_file()
        self.wrapper.change_wallpaper(selected_file,self.config.get_mode())
        self.tray.set_tooltip(os.path.basename(selected_file))
        print "Wallpaper changed to:",selected_file
        return True
       
    def changed_interval(self, *args, **kwargs):
        #print kwargs
        interval = kwargs['time']
        interval_type = kwargs['interval_type']
        #let's be reasonable, m'kay? ;)
        #max 10 hours
        if interval > 0 and interval < self.default_max_interval:
            #if change to preselected
            if interval_type == 0:
                self.config.set_interval(interval)
                #change from custom to preselected 
                if self.config.get_interval_type() == 1:
                    self.tray.deselect_custom_interval_button()
            else:
                self.config.set_custom_interval(interval)
                self.tray.select_custom_interval_button()
                self.tray.set_custom_label("Custom: "+str(self.config.get_custom_interval()))

            self.changed_interval_type(interval_type=interval_type)    
            
    def changed_interval_type(self, interval_type):
        self.config.set_interval_type(interval_type)
        self.stop()
        self.start()
            
    def changed_mode(self,*args,**kwargs):
        mode = kwargs['mode']
        print mode
        if mode in self.wrapper.supported_modes():
            print "Mode changed to",mode
            self.config.set_mode(mode)
            self.change_wallpaper()
             
    def set_folder(self, path):
        self.stop()
        self.fileroller.set_path(path) 
        self.config.set_path(path)
        self.start()

    def start(self,*args,**kwargs):
        #don't start if bogus
        if self.fileroller.has_files() and self.fileroller.is_valid_path():
            self.timer.start()
        else:
            print "Path not valid or no matching files, no files loaded!"
            self.show_select_folder(None,None)
            
    def stop(self,*args,**kwargs):
        self.timer.stop()
        self.tray.set_tooltip("Wally")
    
    def show_about_dialog(self, widget,event):
        about_dialog = WallyAboutDialog()
        about_dialog.run()
        about_dialog.destroy()
    
    def show_select_folder(self, widget,event):
        select_dialog = WallyFolderSelector("Select folder:", self.config.get_path())
        response = select_dialog.run()
        
        if response == gtk.RESPONSE_OK:
            folder = select_dialog.get_filename()
            self.set_folder(folder)
            self.tray.set_path_label(folder)
            
        select_dialog.destroy()    

    def show_select_interval(self, widget,event):
        messagedialog = WallyIntervalInput("Set custom interval (in minutes)")
        response = messagedialog.run() 
        time = messagedialog.get_input()
        messagedialog.destroy()
        
        if response == gtk.RESPONSE_OK:
            self.changed_interval(time=int(time),interval_type=1)
    
                
    def do_quit(self, *args):
        self.stop()
        self.config.save_config()
        gtk.main_quit()
    
    def main(self):
        self.start()

src/WallyWrapper.py
from subprocess import call, check_output, STDOUT, CalledProcessError
import string, os

class WallyWrapper():
    def __init__(self):
        pass
    
    @staticmethod
    def system_has_prerequisites(): 
        if not WallyWrapper.is_installed() or not WallyWrapper.is_managing_desktop() or not WallyWrapper.supports_wallpaper():
            return False
        return True
    
    @staticmethod
    def is_installed():
        try:
            check_output(["which pcmanfm"],stderr=STDOUT,shell=True)
        except CalledProcessError:
            print "PCManFM is NOT installed!"
            return False
        print "PCManFM IS installed!"
        return True
    
    @staticmethod
    def supports_wallpaper():
        output = check_output(["pcmanfm --help | grep wallpaper"],shell=True)
        splitted = string.split(output.strip(), "\n")
        if len(splitted) == 2:
            print "PCManFM SUPPORTS setting wallpaper!"
            return True
        print "PCManFM does NOT support setting wallpaper!"
        return False

    @staticmethod
    def is_managing_desktop():
        output = check_output(["ps ax | grep 'pcmanfm --desktop'"],shell=True)
        splitted = string.split(output.strip(), "\n")
        #print splitted
        if len(splitted) == 3:
            print "PCManFM IS managing desktop!"
            return True
        print "PCManFM is NOT managing desktop!"
        return False
    
    def supported_modes(self):
        output = check_output(["pcmanfm --help | grep wallpaper"],shell=True)
        all_modes = ("stretch","fit","tile","center")
        supported_modes = []
        for mode in all_modes:
            if mode in output:
                supported_modes.append(mode) 
        print "Supports:",supported_modes," modes."
        return supported_modes 
    
    def change_wallpaper(self, file_path, mode):
        #print "pcmanfm --set-wallpaper='"+file_path+"' --wallpaper-mode="+mode
        call(["pcmanfm --set-wallpaper='"+file_path+"' --wallpaper-mode="+mode],shell=True)
    
    @staticmethod
    def user_pictures_folder():
        output = ""
        xdg_installed = True
        
        #for ubuntu and similar
        try:
            output = check_output(["which xdg-user-dir"],stderr=STDOUT,shell=True)
        except CalledProcessError:
            print "xdg-user-dir is NOT installed!"
            xdg_installed = False
        
        if xdg_installed:
            output = check_output(["xdg-user-dir PICTURES"],shell=True)
            output = output.strip()
        #fallback to user home folder
        else:
            output = os.path.expanduser("~")
        
        print "Using",output,"as pictures folder"
        return output

src/WallyFileRoller.py
import os
from random import choice

class WallyFileRoller():
    
    def __init__(self, root):
        self.root = root
        self.files = None
        self.get_files()
    
    def get_files(self):
        if self.is_valid_path():
            matches = []
            for root, dirnames, filenames in os.walk(self.root):
                for filename in filenames:
                    if filename.endswith(('.jpg', '.jpeg', '.gif', '.png', '.bmp', '.tiff')):
                        matches.append(os.path.join(root, filename))
            self.files = matches
            print "Read "+str(len(self.files))+" files"
        else:
            print "Path not valid, no files loaded - using old files, if present!"
            
    def get_next_file(self):
        filepath = choice(self.files)
        return filepath
    
    def set_path(self, path):
        self.root = path
        self.get_files()
    
    def has_files(self):
        if self.files:
            return True
        
        return False
    
    def is_valid_path(self):
        if self.root == "" or not os.path.isdir(self.root):
            return False
        return True
    

src/WallyGUI.py
# -*- coding: utf-8 -*-

import gtk
import os
from functools import partial
from WallyUtil import Plural

class WallyTray(gtk.StatusIcon):
    def __init__(self, manager):
        gtk.StatusIcon.__init__(self)
        self.manager = manager
        self.set_from_file(os.path.dirname(os.path.realpath(__file__))+"/images/tango_desktop_preferences.png")
        self.connect('popup-menu', self.menu_popup)
        self.connect('button-press-event', self.manager.user_change_wallpaper)#self.on_double_click,test=1)
        
        self.set_tooltip('Wally')
        self.menu = None
        
        self.make_menu()
        
    def make_menu(self):
        self.menu = gtk.Menu()
        
        self.create_quit_item()
        self.create_about_item()
        
        #separator
        sep = gtk.SeparatorMenuItem()
        sep.show()
        self.menu.append(sep)
       
        self.create_path_items()
        self.create_mode_items()
        self.create_interval_items()
        
        #separator
        sep = gtk.SeparatorMenuItem()
        sep.show()
        self.menu.append(sep)
       
        self.create_stop_item()
        self.create_start_item()

    
    def create_quit_item(self):
        # add quit item
        self.quit = gtk.MenuItem("Quit")
        self.quit.show()
        self.quit.connect('button-press-event', self.manager.do_quit)
    
        self.menu.append(self.quit)
        
    def create_about_item(self):
        about = gtk.MenuItem("About")
        about.show()
        about.connect('button-press-event', self.manager.show_about_dialog)

        self.menu.append(about)
    
    def create_path_items(self):
        path = gtk.MenuItem("Path")
        path.show()
        self.menu.append(path)

        #path submenu
        self.path_submenu = gtk.Menu()
        
        if not self.manager.fileroller.has_files() or not self.manager.fileroller.is_valid_path():
            set_path = "Invalid path, select new"
        else:
            set_path = self.manager.config.get_path()
            
        self.path_real = gtk.MenuItem(set_path)
        self.path_real.show()
        path_sep = gtk.SeparatorMenuItem()
        path_sep.show()
        path_open = gtk.MenuItem("Select...")
        path_open.show()
        path_open.connect('button-press-event', self.manager.show_select_folder)
          
        self.path_submenu.append(self.path_real)
        self.path_submenu.append(path_sep)
        self.path_submenu.append(path_open)
        
        path.set_submenu(self.path_submenu)
    
    def create_mode_items(self):
        #modes
        modes = gtk.MenuItem("Mode")
        modes.show()
        self.menu.append(modes)
        
        #modes submenu
        self.interval_modes = gtk.Menu()
        set_mode = self.manager.config.get_mode()
        
        self.mode_buttons = []
        for mode in self.manager.wrapper.supported_modes():# ("Stretch","Fit","Center","Tile"):
            if mode == "stretch":
                group = None
            else:
                group = self.mode_buttons[0]
                 
            button = gtk.RadioMenuItem(group,mode.title())
            button.connect('button-press-event', partial( self.manager.changed_mode,mode=mode))
            button.show()
            if set_mode == mode:
                button.set_active(True)
                
            self.mode_buttons.append(button)
            self.interval_modes.append(button)
    
    
        modes.set_submenu(self.interval_modes)

    def create_interval_items(self):
        interval = gtk.MenuItem("Interval")
        interval.show()
        
        #interval values submenu
        self.interval_submenu = gtk.Menu()
        
        set_interval = self.manager.config.get_selected_interval()
        set_interval_type = self.manager.config.get_interval_type()
        
        self.buttons = []
        
        #custom interval
        self.custom = gtk.RadioMenuItem(None,"Custom: "+str(self.manager.config.get_custom_interval()))
        self.custom.show()
        self.custom.connect('button-press-event', partial(self.manager.changed_interval,time=self.manager.config.get_custom_interval(),interval_type=1))
        self.custom.connect('button-release-event',self.disable_propagation)
        
        self.interval_submenu.append(self.custom)
        self.buttons.append(self.custom)
        
        self.custom_interval = gtk.MenuItem("Change")
        self.custom_interval.show()
        self.custom_interval.connect('button-press-event', self.manager.show_select_interval)
        
        if set_interval_type == 1:
            self.custom.set_active(True)
        else:
            self.custom_interval.set_sensitive(False)
        
        self.interval_submenu.append(self.custom_interval)
        
        #preselected intervals
              
        self.custom_interval_sep = gtk.SeparatorMenuItem()
        self.custom_interval_sep.show()
        self.interval_submenu.append(self.custom_interval_sep)
         
        group = self.buttons[0]
        for time in self.manager.default_intervals:
            button = gtk.RadioMenuItem(group,Plural(time,"minute"))
            #button.connect('button-press-event',self.radio_button_activated,time*60)
            button.connect('button-press-event',partial(self.manager.changed_interval,time=time,interval_type=0))

            button.show()
            if set_interval == time and set_interval_type == 0:
                button.set_active(True)
                
            self.buttons.append(button)
            self.interval_submenu.append(button)
    
    
        interval.set_submenu(self.interval_submenu)

        self.menu.append(interval)
    
    def create_stop_item(self):
        self.stop = gtk.MenuItem("Stop")
        self.stop.show()
        self.menu.append(self.stop)
        self.stop.connect('button-press-event', self.manager.stop)
       
    def create_start_item(self):
        self.start = gtk.MenuItem("Start")
        self.start.show()
        self.menu.append(self.start)
        self.start.connect('button-press-event', self.manager.start)
            
    def disable_propagation(self, widget, event, **kwargs):
        #to prevent menu from closing on menu item click        
        return True
    
    def menu_popup(self, icon, event_button, event_time):
        self.menu.popup(None, None, gtk.status_icon_position_menu, event_button, event_time, self)
        self.menu.show()
    
    def deselect_custom_interval_button(self):
        self.custom.set_active(False)
        self.custom_interval.set_sensitive(False)
    
    def select_custom_interval_button(self):
        self.custom.set_active(True)
        self.custom_interval.set_sensitive(True)
        
    def set_custom_label(self, label):
        self.custom.set_label(label)
    
    def set_path_label(self, label):
        self.path_real.set_label(label)
            
class WallyFolderSelector(gtk.FileChooserDialog):
    def __init__(self, message,folder):
        gtk.FileChooserDialog.__init__(self,message,None,gtk.FILE_CHOOSER_ACTION_SELECT_FOLDER,(gtk.STOCK_CANCEL,gtk.RESPONSE_CANCEL,gtk.STOCK_OPEN,gtk.RESPONSE_OK))
        self.set_current_folder(folder)
        self.set_default_response(gtk.RESPONSE_CANCEL)
        
class WallyIntervalInput(gtk.MessageDialog):
    def __init__(self, message):
        gtk.MessageDialog.__init__(self,parent=None,flags=0,type=gtk.MESSAGE_QUESTION, buttons=gtk.BUTTONS_OK_CANCEL, message_format=message)
        
        action_area = self.get_content_area()
        adj = gtk.Adjustment(1, 0, 600, 1, 1)
        self.entry = gtk.SpinButton(adj)
        action_area.pack_start(self.entry)

        self.show_all()
        
    def get_input(self):
        time = self.entry.get_value_as_int()
        return time
        
class WallyAboutDialog(gtk.AboutDialog):
    def __init__(self, *args, **kwargs):
        gtk.AboutDialog.__init__(self, *args, **kwargs)
        self.set_destroy_with_parent (True)
        self.set_icon_name ("Wally")
        self.set_name('Wally')
        self.set_version('0.1 alpha')
        self.set_copyright("(C) 2013 paxman")
        self.set_comments(("Desktop background changer for PCManFM file browser"))
        
class WallyMessageDialog(gtk.MessageDialog):
    def __init__(self, message):
        gtk.MessageDialog.__init__(self,parent=None, flags=0, type=gtk.MESSAGE_ERROR, buttons=gtk.BUTTONS_OK, message_format=message)
        self.run()
        self.destroy()
          
        